fenics/environment.yml

```diff
@@ -0,0 +1,9 @@
+channels:
+- conda-forge
+dependencies:
+  - fenics =2019.1.0
+  - fenics-libdolfin =2019.1.0=*_8
+  - mshr =2019.1.0=*_2
+  - pyiron =0.3.21
```

Comment:
Do you need anything specific from `pyiron`? or is `pyiron_base` sufficient? To my understanding `fenics` should only be based on `pyiron_base`.

Reply:
The issue is that we then have to update `pyiron` at multiple places which might be error prone. So I would prefer to only use `pyiron_base` here.

Reply:
> 
> 
> The issue is that we then have to update `pyiron` at multiple places which might be error prone. So I would prefer to only use `pyiron_base` here.

I see your point. I am trying to see whether I can run Liam's notebook with importing Projects from pyiron_base instead of pyiron. If it is successful, we can drop pyiron.

Reply:
Unfortunately, it fails when `job.create.domain` is used, as `job` does not have `create` as an attribute.
As a solution to avoid having the need to specify pyiron version in multiple places, we can use `pyiron/md` as the base image for `pyiron/fenics`. I know it will increase the size of the image, but we did also this for `pyiron/damask` image as well.

Reply:
@liamhuber Can you comment on this? Is `pyiron` required for `FEniCS`? 

@muh-hassani The difference with `damask` is that in the example we couple `Lammps` and `Damask` so using `pyiron/md` makes a lot of sense.

Reply:
> @liamhuber Can you comment on this? Is `pyiron` required for `FEniCS`?

@jan-janssen, pyiron (atomistic) is not required. So far the only pyiron-dependencies are `GenericJob`, `InputList`, and `PyironFactory`, all from base.

> Unfortunately, it fails when `job.create.domain` is used, as `job` does not have `create` as an attribute.

@muh-hassani I'm surprised by this. The `Fenics` class inherits explicitly sets `create` in `__init__`. The creator in turn has `domain` and `bc` (boundary condition) attributes for those respective factories. What is the `type` of the job that's giving you the error?

> As a solution to avoid having the need to specify pyiron version in multiple places, we can use `pyiron/md` as the base image for `pyiron/fenics`. I know it will increase the size of the image, but we did also this for `pyiron/damask` image as well.

Yes, I would like this solution. For the lesson in January I don't *currently* have a plan to couple directly to atomistics, but it's a nice option to have since, e.g., the elastic matrix is good input for fenics just like it is for DAMASK.

Reply:
@liamhuber, sorry. My comment about the `job.create.domain` was wrong. The error is related to `pr.create.job.Fenics()`
The code is as follows:
```
from pyiron_base import Project
import pyiron_contrib
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pr = Project('fenics_tutorial')
pr.remove_jobs_silently(recursive=True)
job = pr.create.job.Fenics('poisson', delete_existing_job=True)
job.domain = job.create.domain.unit_mesh.square(8, 8)
```
It produces the following error:
```
AttributeError: 'Project' object has no attribute 'create'
```

Reply:
> It produces the following error:
> 
> ```
> AttributeError: 'Project' object has no attribute 'create'
> ```

Ah, that's fair enough -- it must be leaning on an outdated version of pyiron_base, the most recent tag has such an attribute.

Reply:
In the mean time I finally got access to the `fenics` and `mshr` conda packages. Starting from Build 4 for `mshr` and build 18 for `fenics` the installation should again work just using: 
```
conda install -c conda-forge mshr
```
But I still need to do some testing and there is always the risk that based on some dependencies already installed on the system it can resolve to older build numbers. So we still need some more testing and some more people getting involved in the conda package management.

Reply:
@jan-janssen how shall we proceed now with pyiron dependency for FEniCS? 
Although the size of the image would be larger, but I am in favor of basing the image on pyiron/md as it provides the chance for combining LAMMPS and FEniCS in future.

Reply:
For now I would base it on `pyiron_base` as @liamhuber said it should work just fine, we can increase the image later any way, but for the moment I guess it helps us to strengthen the point that pyiron does not require the atomistic component.

Reply:
I tried to modify Liam's notebook to use pyiron_base. But even by importing pyiron_contrib, `pyiron` is required. The error I get is that in `pyiron_contrib/__init__.py`, the `Project` is imported from `pyiron`:
```
/pyiron_contrib/__init__.py in <module>
      2 __all__ = []
      3 
----> 4 from pyiron import Project
      5 from pyiron_base import JOB_CLASS_DICT
      6 

ModuleNotFoundError: No module named 'pyiron'
```
If we want to use the examples from pyiron_contrib, there should be a modification either in pyiron_contrib to use pyiron_base instead of pyiron or  here in the `environment.yml`.

Reply:
@muh-hassani yes, using `pyiron_base` in my notebooks is 100% fine. Sorry about that, it was just muscle-memory that made it `pyiron`. I fixed it.

Re `pyiron_contrib/__init__.py`, I'll go peek at the blame.

Reply:
Ok, as far as I understand this line allows people to say `from pyiron_contrib import Project`, which, because we add to the `JOB_CLASS_DICT` further down in the file, is equivalent to `from pyiron import Project; import pyiron_contrib`.

If this understanding is correct, I say we just get rid of that import line. People's notebooks might fail if they're using the first syntax, but since it's contrib they can deal with it. Raynol's example Protocol notebooks in the repo already use my preferred syntax `from pyiron import Project; import pyiron_contrib`.

Reply:
@liamhuber thanks for the update to the notebook. So I will remove this import line and push it to the branch.

Reply:
> Ok, as far as I understand this line allows people to say `from pyiron_contrib import Project`, which, because we add to the `JOB_CLASS_DICT` further down in the file, is equivalent to `from pyiron import Project; import pyiron_contrib`.
> 
> If this understanding is correct, I say we just get rid of that import line. People's notebooks might fail if they're using the first syntax, but since it's contrib they can deal with it. Raynol's example Protocol notebooks in the repo already use my preferred syntax `from pyiron import Project; import pyiron_contrib`.

I would suggest to do it with a try except statement, if pyiron is available it can be used if it is not then fall back to pyiron_base. In future we can replace the try except part with a function which checks which pyiron modules are available and returns the corresponding project.

Reply:
Now, pyiron is removed from the list of the dependencies.